addons/l10n_eg_edi_eta/models/eta_thumb_drive.py

```diff
@@ -0,0 +1,166 @@
+from odoo.exceptions import ValidationError
+
+
+class EtaThumbDrive(models.Model):
+    _name = 'l10n_eg_edi.thumb.drive'
+    _description = 'Thumb drive used to sign invoices in Egypt'
+
+    user_id = fields.Many2one('res.users', required=True, default=lambda self: self.env.user)
+    company_id = fields.Many2one('res.company', required=True, default=lambda self: self.env.company)
+    certificate = fields.Binary('ETA Certificate')
+    pin = fields.Char('ETA USB Pin', required=True)
+
+    _sql_constraints = [
+        ('user_drive_uniq', 'unique (user_id, company_id)', 'You can only have one thumb drive per user per company!'),
+    ]
+
+    def _create_token(self, invoice_ids, url, max_ts):
+        self.ensure_one()
+        msg = '%s%s%s%s%s' % (url, invoice_ids, self.env.user.id, self.id, max_ts)
+
+        secret = self.env['ir.config_parameter'].sudo().get_param('database.secret')
+        assert secret, "CSRF protection requires a configured database secret"
+        return hmac.new(secret.encode('ascii'), msg.encode('utf-8'), hashlib.sha256).hexdigest()
+
+    def action_sign_invoice(self, invoice_ids):
+        self.ensure_one()
+        sign_host = self._get_host()
+
+        url = self.get_base_url()
+        max_ts = int(time.time() + 120)  # valid for two minute
+        hm = self._create_token(invoice_ids, url, max_ts)
+        return {
+            'type': 'ir.actions.client',
+            'tag': 'action_post_sign_invoice',
+            'params': {
+                'sign_host': sign_host,
+                'otp': '%so%s' % (hm, max_ts),
+                'url': url,
+                'user': self.env.user.id,
+                'pin': self.pin,
+                'drive': self.id,
+                'invoices': invoice_ids
+            }
+        }
+
+    def action_set_certificate_from_usb(self):
+        self.ensure_one()
+        sign_host = self._get_host()
+
+        url = self.get_base_url()
+        max_ts = int(time.time() + 120)  # valid for two minute
+        hm = self._create_token(False, url, max_ts)
+
+        return {
+            'type': 'ir.actions.client',
+            'tag': 'action_get_drive_certificate',
+            'params': {
+                'sign_host': sign_host,
+                'otp': '%so%s' % (hm, max_ts),
+                'url': url,
+                'user': self.env.user.id,
+                'pin': self.pin,
+                'drive': self.id
+            }
+        }
+
+    def _get_host(self):
+        sign_host = self.env['ir.config_parameter'].sudo().get_param('default.sign.host')
+        if not sign_host:
+            raise ValidationError(_('Please define the host of sign tool.'))
+        return sign_host
+
+    def _serialize(self, eta_inv):
+        # https://sdk.preprod.invoicing.eta.gov.eg/document-serialization-approach/
+        if isinstance(eta_inv, dict):
+            return json.dumps(eta_inv)
+
+        canonical_str = ''
+        for key, value in eta_inv.items():
```

Comment:
I think it should respect the original order.  Is not the dict ordered now? @xmo-odoo

Reply:
Dicts keep insertion order, and json loads and json dumps are also supposed to preserve it (unless told otherwise).

That doesn't mean it's what the signing format wants.